Replace Nullable{T} with Union{Some{T}, Void}

nalimilan opened the pull request:
This is a first proposal to fix #22682. Tests pass but I haven't updated docs and the `broadcast` code needs checking.  What this PR does currently is:
- Deprecate `Nullable`, moved to a new [Nullables.jl](https://github.com/nalimilan/Nullables.jl) package
- Introduce `Null`/`null` from Nulls.jl, `Some{T}` and `Optional{T} = Union{Null, Some{T}}` as a replacement for `Nullable` in cases where either 1) one wants to force explicit unwrapping and/or 2) one needs to allow wrapping `null` inside an `Option` and be able to distinguish it from a `null` (typically, for `tryget` on dictionaries or `tryparse` on a nullable value). `Union{T, Null}` is still available to represent missing values, and will allow silent propagation of nulls once we implement operators on `Null`.
- Rename `get(::Nullable[, y])` to `unwrap(::Some[, y])`: this is not strictly needed, but allows seeing to extent to which it's used in Base, and evaluating how simpler the code would be if we used `Union{Null, T}` rather than `Union{Null, Some{T}}` (and therefore didn't require unwrapping)

Issues to discuss include:
- Naming: `Option` (Rust, Scala, ML, F#) could be `Optional` (Swift, C++17, Java 8) or `Maybe` (Haskell). It seems that the two former are [the most common](https://en.wikipedia.org/wiki/Option_type). `Some` (used by Rust, Scala, ML, F#, Swift) could also be `Value` or `Just` (as in Haskell).
- Whether to support `convert(Some{T}, ::T)`. At first I thought it was convenient (we allowed it for `Nullable`), but I realized it wasn't a great idea since it implies that `convert(Option{T}, x)` gives `Some(x)` in general but `null` when `x = null`, which is a trap transforming a wrapped `null` (which should be `Some(null)`) into a null `Option`, defeating the purpose of using a wrapper in the first place.
- What to do with `broadcast`: things like `broadcast(exp, Some(1))` work, but an error is thrown for `broadcast(exp, null)`, which makes this operation mostly useless. We can fix this by defining operators on `Null` (just like in Nulls.jl), but that won't work for all functions. We should be able to add a special rule for `Null` to the `broadcast` machinery, that's probably worth it since that's one of the main features of `Option` in many languages.
- Whether to introduce another form of missing value for `Option` in addition to `Void`/`nothing` and `Null`/`null`, which could be `None`/`none` for consistency with other languages. The main argument in favor of this would be to distinguish clearly `Option` from `?T`/`Union{Null, T}`, since the latter is intended for missing values in data with automatic propagation of nulls. But I think most people were opposed to this in the discussion.

nalimilan committed:
Replace Nullable{T} with Union{Some{T}, Void}

nalimilan commented on base/null.jl:
```diff
@@ -0,0 +1,116 @@
+
+convert(::Type{Option{T}}, ::Null) where {T} = null
+convert(::Type{Option   }, ::Null)           = null
+# FIXME: find out why removing these two methods makes addprocs() fail
```
This has turned out to be very hard to debug. Without these two methods, `addprocs(2)` hangs and finally times out, leaving two julia processes using 100% CPU in the background. Replacing all occurrences of `nothing` with `null` in base/distributed/* didn't fix it. The output isn't very helpful. Any help on how to debug this is welcome!

```julia
julia> addprocs(2)
ERROR: Timed out waiting to read host:port string from worker.
read_worker_host_port(::Pipe) at ./distributed/cluster.jl:302
connect(::Base.Distributed.LocalManager, ::Int64, ::WorkerConfig) at ./distributed/managers.jl:395
create_worker(::Base.Distributed.LocalManager, ::WorkerConfig) at ./distributed/cluster.jl:507
setup_launched_worker(::Base.Distributed.LocalManager, ::WorkerConfig, ::Array{Int64,1}) at ./distributed/cluster.jl:453
(::getfield(Base.Distributed, Symbol("##41#44")){Base.Distributed.LocalManager,WorkerConfig})() at ./task.jl:335

...and 1 more exception(s).

Stacktrace:
 [1] sync_end() at ./task.jl:287
 [2] macro expansion at ./task.jl:303 [inlined]
 [3] #addprocs_locked#38(::Array{Any,1}, ::Function, ::Base.Distributed.LocalManager) at ./distributed/cluster.jl:407
 [4] #addprocs#37(::Array{Any,1}, ::Function, ::Base.Distributed.LocalManager) at ./distributed/cluster.jl:371
 [5] #addprocs#257(::Bool, ::Array{Any,1}, ::Function, ::Int64) at ./distributed/managers.jl:314
 [6] addprocs(::Int64) at ./distributed/managers.jl:313
```

quinnj replied:
Perhaps related to automatic conversion thru fallback constructor? https://github.com/JuliaLang/julia/pull/23273

nalimilan replied:
Generally this rather happens when assigning to an `Option` field. It's quite rare to call `Option(x)`.

I've finally found a way to see the backtrace from the workers, by adding print statements to `read_worker_host_port`. Turns out the `TCPSocket` inner constructor used `nothing` to set a `Nullable` field to null. Now it works!

nalimilan commented on test/choosetests.jl:
```diff
@@ -30,12 +30,14 @@ function choosetests(choices = [])
+        "boundscheck", "error", "cartesian", "asmvariant", "osutils",
+        "channels", "iostream", "specificity", "codegen", "codevalidation",
+        # FIXME: fix ambiguities
+        "ambiguous"
```
This test fails due to ambiguities between the following methods:
```julia
convert(::Type{Union{Null, Some{T}}}, x::Some) where T
convert(::Type{Union{Null, Some{T}}}, ::Void) where T
convert(::Type{Union{Null, Some{T}}}, ::Null) where T
```

How can I fix this? My attempts have not been successful so far.

StefanKarpinski commented:
Is `Some` a standard name for this immutable wrapper type? We had also discussed `Value`, but that might be confusing with `Val` (which is also a confusing name, but that's another issue).

ararslan commented:
Thanks so much for working on this! It's exciting to see this finally coming to fruition.

My take on the issues you've enumerated:

**Naming**:
* +1 for `Some` over `Value`, since `Value` is too similar to `Val`.
* +1 for `Maybe` over `Option` and `Optional`. `Option` would conflict with #23637 and in my opinion `Optional` isn't terribly descriptive. To me, `Maybe{T}` says "maybe this is some `T` and maybe it's `null`."

**Supporting conversion to `Some{T}`**:
* I don't see why `convert(Some{T}, x)` implies that `convert(Option{T}, x)` should work. In my mind, it would only make sense to allow conversion to `Option{T}` if _both_ conversion to `Some{T}` _and_ conversion to `Null` were allowed, but I don't think it makes sense to allow conversion to `Null` in the general case (just the no-op `convert(Null, null)`).

**Broadcasting**:
* I don't think we should define `broadcast`/`map` to do unwraping of any kind. Those functions are for vectorization and frankly I think it was a bad decision to use `broadcast` for `Nullable`.
* `broadcast(exp, null)` should just be `exp(null)` since `null` is a scalar, and we should define `exp(null)` to be `null`. The same should be done for many other Base functions as a way to indicate that they know how to deal with missing values.

**Yet another type**:
* I don't understand why that would be necessary. -1 for adding a `None`/`none`. Seems like we have our bases covered with `Union{T, Null}` and `Union{Some{T}, Null}`.

ararslan commented:
> Is `Some` a standard name for this immutable wrapper type?

As one data point it's what Rust uses. Not sure what else uses it.

nalimilan commented:
Yes, according to [Wikipedia](https://en.wikipedia.org/wiki/Option_type), `Some` is used by Rust and Scala (Swift uses `some`). Another common term is `Just` (in Haskell notably).

quinnj commented on base/null.jl:
```diff
@@ -0,0 +1,116 @@
+an `Option` object is null, and [`unwrap`](@ref) to access the value inside the `Some`
+object if not.
+"""
+Option{T} = Union{Null, Some{T}}
```
This doesn't need a `const`, right?

nalimilan commented:
> Naming:
> 
>     +1 for Some over Value, since Value is too similar to Val.
>     +1 for Maybe over Option and Optional. Option would conflict with #23637 and in my opinion Optional isn't terribly descriptive. To me, Maybe{T} says "maybe this is some T and maybe it's null."

Ah, good point about #23637. Though the issue with `Maybe` is that it seems languages that have it (like Haskell) use `Nothing` or `nothing` to denote no value. That could help confusion with our `nothing`. OTOH languages which have `Option` or `Optional` often use `None`, which differs from our `Null` but at least does not exist at all in the language. So I guess I'd prefer `Optional`.

> Supporting conversion to Some{T}:
> 
> I don't see why convert(Some{T}, x) implies that convert(Option{T}, x) should work. In my mind, it would only make sense to allow conversion to Option{T} if both conversion to Some{T} and conversion to Null were allowed, but I don't think it makes sense to allow conversion to Null in the general case (just the no-op convert(Null, null)).

Why not, but there are probably very few cases in which one would like to convert from `T` to `Some{T}`. The most common case (see the diff) is to assign to an `Option{T}` field. So basically the decisive issue is whether or not to allow conversion to `Option`.

> Broadcasting:
> 
>     I don't think we should define broadcast/map to do unwraping of any kind. Those functions are for vectorization and frankly I think it was a bad decision to use broadcast for Nullable.

Yet in most languages supporting `Option`, the recommended way of working with them is to use operations on collections like `map`/`broadcast`, considering them as zero- or one-element containers. See for example [the use of `flatMap` in Scala](http://danielwestheide.com/blog/2012/12/19/the-neophytes-guide-to-scala-part-5-the-option-type.html). I wasn't a fan of treating `Nullable` as a container, but now that it's clear missing data should be represented using `Union{T, Null}`, I don't have any problem with treating `Option` like that. Cf. also the long discussion for `Nullable` at https://github.com/JuliaLang/julia/pull/16961.

Anyway, we can always start without this and add it later if needed. That feature doesn't seem to be used in Base, which might indicate that it's not essential (but be sure people will ask for it). But I'll have to actively remove code to prevent it from working at all, since the basics already exist due to the `Nullable` legacy.

rfourquet commented:
Great! I didn't follow closely this switch-away-from-nullable effort, so I don't see why we wouldn't have `Optional{T} = Union{Void, Some{T}}` ? I understand that `Null` is needed for `Union{Null, T}`, because `T` could be `Void` (I think?), but for optional we might as well go the simpler option with `Void`. After all, assuming `nothing` is a must-know in Julia, someone could then learn `Optional` with minimal overhead; `Null` would be only for the "data scientist working with missing value". And isn't it an advantage to separate completely those two types which were before merged in `Nullable` ? (genuine question)

ararslan commented:
The problem with using `Void` is that it gives the type a different meaning. `nothing` is what's returned from loops, `print`, and other things like that, so it would be weird to say that if a value isn't present, it's the same as what you'd get from assigning to the result of `print`.

nalimilan commented:
@rfourquet We discussed that a bit in #22682. There's no intrinsic reason to use `Null` rather than `Void` here, since `T` could equally be `Null` or `Void` (the solution to that is to use `Union{Null, Some{T}}` or `Union{Void, Some{T}}`, see above). It's mostly a matter of convention. Since `null` is already meant to denote the absence of a value, it sounds logical to use it. We need `isnull` and `get`/`unwrap` anyway, so better use it both for `Option{T}` and for `Union{Null, T}`. That way we can keep `nothing` for its only documented use, that is as the return value of blocks which do not return anything.

That said, one small drawback of reusing `Null` here is that we plan to define things like `exp(::Null) = null` to allow for propagation of missing values. That means that `exp(::Option{T})` would throw an error when a value is present (i.e. with a `Some{T}` argument), but would succeed when the value is missing (`null` argument). That's counter-intuitive, though in practice it's probably not a big deal since people generally test their code with valid values -- what they forget to test is when the value is missing.

jamesonquinn commented:
I like all the general ideas here. 

My somewhat non-bikeshedding comment: we should definitely have different values for "or maybe not" (that is, `Union{Some{T}, XXX}`) and "na" (that is `?T = Union{T, YYY}`). The whole point of these two objects is that they have different semantics.

My purely bikeshedding comments:

For `?T = Union{T, YYY}` (data scientist's null), I think YYY should be `na::NA`, after R. Relatively well-known and well-understood, name explains pretty much what it means; data that's not available.

For `Union{Some{T}, XXX}` (computer scientist's null), I think XXX should be either `nothing::Void` (because why multiply our null-like values for no reason), or if not, either `empty::Empty` or `null::Null`.

TotalVerb commented:
My two cents:

- No conversion from `T` to `Some{T}` is needed. This was useful in cases where you have a field/array of a known type that is nullable, but those cases can be replaced by `Union{T, Null}` instead of `Option{T}`. Cases where wrapped nulls may be necessary always require explicit wrapping instead of implicit conversion to be correct.
- Shouldn't we define `broadcast(f, ::Null) = null`, which seems to be simple and correct?

TotalVerb commented:
Another option is `Union{Tuple{T}, Tuple{}}` which gives covariance and the `broadcast` behavior for free. But this is perhaps a little too much of a pun.

nalimilan commented:
@jamesonquinn The ship has sailed on using `null` for missing values. We're not going to change DataFrames to use `na` at this point.

> My somewhat non-bikeshedding comment: we should definitely have different values for "or maybe not" (that is, Union{Some{T}, XXX}) and "na" (that is ?T = Union{T, YYY}). The whole point of these two objects is that they have different semantics. For instance, == and === should work the same for the former, not for the latter.

What do you mean? What exact behavior would you expect these operators would do in each case?

nalimilan commented:
> Shouldn't we define broadcast(f, ::Null) = null, which seems to be simple and correct?

@TotalVerb That wouldn't help with more complex cases like `x .+ y`, right? We need something more general if we want to support `broadcast` on `Option`.

rfourquet commented:
Thanks for your answers. So I read a bit more the threads, in which some of my impressions were already expressed: 

1. That we want it or not, I believe `Union{T, Void}` will be a de facto goto solution for software-null: it's already used (even in base, maybe from a time predating `Nullable`), is extremly simple, it became efficient, and most of the time there is no ambiguity on the meaning of `nothing` (i.e. it's not a possible value of `T`). 
2. I didn't find a clear comparison of the merits of `Union{Some{T}, None}` (let's call it `None` here to think independently of data-null)  vs a `Maybe` type which could be implemented exactly like `Nullable`, or alternatively:
```julia
struct Maybe{T}
    value::Union{Some{T}, None}
    Maybe{T}() where T = new(None())
    Maybe{T}(x) where T = new(Some(x))
end
``` 
I can't come up with solid arguments for now, but I feel that it's my prefered solution:
+ It allows to deal with a well specified type, `Maybe`, which is the only new introduced name (`Some` remains an implementation detail). Otherwise, I'm pretty sure we will have `const Maybe{T} = Union{Some{T}, None} where T`, which makes it two new names. Without counting `None`, if it's different from `Void`, and from `Null`...
+ it seems better for type stability; even though performance would not be a concern anymore in this case, type stability helps reasoning about the code; I also imagine having a typed absent-value can come in handy.  
+ `Some{T}` considered independently of the `Union` appears strange and meaningless (imagine finding a function definifion like `f(x::Some{Int}) = ... `). Also you can not prevent someone from using `Union{Some{T}, Void}` or even `Union{Some{T}, CustomNone}`. Given point 1, `Union{Some{T}, Void}` seems to offer very little value over `Union{T, Void}` when no ambiguity is possible. I guess my point is that fragmentation of usages can't be avoided with the `Union{Some{T}, None}` solution. With `Maybe{T}`, it's not possible to split one component out of the "context of optional value". 
+ `Some` being an implementation detail, `None` also is, and can as well be `Void` after all, but it doesn't matter.

PS: note that I saw [e.g here](https://github.com/JuliaLang/julia/issues/22682#issuecomment-313106169) the similar proposition `struct Maybe{T}; x::Union{T, None}; end`, but which doesn't allow to discriminate an "absent value" from a "present `None()`"; I guess it's an simple oversight as it was presented as an alternative implementation of `Nullable`.

nalimilan commented:
@rfourquet I don't understand the point of defining `Maybe{T}` as you do above. Using `Some` inside a `struct` is redundant (you wrap values twice). You seem to be arguing in favor of keeping the current `Nullable` instead. That approach shares the same limitations as `Nullable`, in particular the "counter-factual `T`" issue (i.e. you need to compute `T` even when you don't have any value, which is sometimes problematic, cf. #22682).

Regarding the fact that people could define their custom `Union{CustomNull, Some{T}}`, I really don't see why they would do that when we already provide `Null` for that. They could be tempted to use `Void`, but `isnull` wouldn't work for it. Finally, in some cases using a different type instead of `Null` could be useful, e.g. to implement a `Result` type of to distinguish multiple kind of nulls.

rfourquet commented:
I interpreted the last paragraph of [this comment](https://github.com/JuliaData/Nulls.jl/issues/6#issuecomment-306086548) as meaning that the `Union` representation can be more efficient than the `Nullable` implementation, hence the idea to wrap the `Union` to replace `Nullable`, but I guess this is irrelevant to the design question and I should have refrained from going into this detail: in other words, I was indeed arguing in favor of a rebranded `Nullable`. 

I was not aware of the "counterfactual return type problem" (presented [there](https://github.com/JuliaLang/Juleps/blob/jmw/nullable/Nullable.md)), and I agree that it's quite convincing.

jamesonquinn commented:
@nalimilan Using `na` as the DS null and `none` as the CS one, purely for purposes of illustration, I'd expect `na == na` to return `na` (just as `3. * na` would), while `none == none` and `na === na` would both be `true`.

This is merely one of many examples where I think the semantics of these values would be different.

nalimilan commented:
> @nalimilan Using na as the DS null and none as the CS one, purely for purposes of illustration, I'd expect na == na to return na (just as 3. * na would), while none == none and na === na (and NA == NA, that is, the type) would both be true.

OK, I see. I wanted this behavior (three-valued logic) too for a long time, see https://github.com/JuliaStats/NullableArrays.jl/pull/85 for a long discussion about that.  Unfortunately, it turned out to be quite inconvenient, since Julia expects `Bool` in many places. After moving away from `Nullable{Bool}` to `Union{Null, Bool}`, we might be able to revisit this decision since at least all operations expecting a `Bool` would work in the absence of nulls even if `null == null` returned `null`; not sure. Anyway, we should have this discussion in a different issue I think, as it tends to generate long debates.

nalimilan commented:
BTW, I've just created the [Nullables.jl](https://github.com/nalimilan/Nullables.jl) package by taking all code and tests from Base. It works on Julia 0.6. and 0.7. Comments welcome (please file issues there), at some point it should probably be moved to JuliaArchive.

ararslan commented:
> Shouldn't we define `broadcast(f, ::Null) = null`, which seems to be simple and correct?

No, that's incorrect for things like `isnull`. Yes, we can specialize on `broadcast(::typeof(isnull), ::Null)`, but there are likely other examples that would come up and have a subtly wrong fallback.

nalimilan commented:
I think this `broadcast` definition is fine for `isnull` too. `isnull.(Some(1))` should return  `Some(false)` and `isnull.(Some(null))` should return `true`, but `isnull.(null)` should return `null`, i.e. propagate missingness. That's very different from calling `isnull(null)`, where you want to test whether the `Option` itself is null: when broadcasting, you want to apply the function to the wrapped value, or return `null` if there's no value. This is actually equivalent to how it works for `Nullable` currently.

Anyway, the discussion about `broadcast` would better be separated, since we are not required to support it with `Option`.

ararslan commented:
I'd prefer we not support `broadcast` with `Option(al)`/`Maybe` anyway.

nalimilan commented:
I've managed to fix the bug I mentioned above. In order to make progress on this PR, I've added commits to simplify the PR:
- Do not introduce the `Option`/`Optional`/`Maybe` type alias (at least for now), but use directly `Union{Some{T}, Null}` instead. As @StefanKarpinski noted, we will be able to write it as `Some{T}?` in 1.0, which is quite short.
- Use `get` as for `Nullable` rather than `unwrap`. The diff is much smaller, but individual commits still allow seeing how commonly this function is used (which was the point of the rename).
- Disable `broadcast` support on `Some` for now. That question is quite complex and tends to generate long debates, so better tackle it in a subsequent issue/PR.

Tests pass, but Documenter will fail until it's updated to use Nullables.jl. However, as a good indication, the FreeBSD CI passes.

Is everybody OK with the new design? It remains to be decided whether to use `Some{T}` or something else, like `Value{T}` -- keeping in mind that it will most commonly be written as `Some{T}?` or `Value{T}?`. `Value` isn't used by any language which has optional/maybe types, but since the PR no longer uses that terminology, that might be OK, or even an advantage.

I'll update the docs once the design is settled.

TotalVerb commented:
While it's the current name, I don't think `get` is the right now. Every other method of `get` require a default value to be passed.

nalimilan commented on base/parse.jl:
```diff
@@ -216,22 +215,39 @@ end
 tryparse(::Type{T}, s::AbstractString) where {T<:Union{Float32,Float64}} = tryparse(T, String(s))
 
-tryparse(::Type{Float16}, s::AbstractString) = convert(Nullable{Float16}, tryparse(Float32, s))
+function tryparse(::Type{Float16}, s::AbstractString)
```
The replacement is much more verbose than the original because the shorter form:
```julia
convert(Union{Some{Float16}, Null}, tryparse(Float32, s))
```
does not work without a `convert` method like this:
```julia
convert(::Type{Union{Some{T}, Null}}, x::Some) where {T} = convert(Some{T}, x)
```

However, adding that `convert` method triggers a failure in the `ambiguous` test:
```julia
detect_unbound_args in Base and Core: Test Failed
  Expression: need_to_handle_undef_sparam == Set()
   Evaluated: Set(Method[convert(::Type{Union{Null, Some{T}}}, x::Some) where T in Main at REPL[1]:1]) == Set{Any}()
```

Does this indicate a real problem with that `convert` method? Or is it OK to add it to the already present exceptions in the test?

TotalVerb replied:
`Nulls.jl` defines

```julia
Base.convert(::Type{Union{T, Null}}, x) where {T} = convert(T, x)
```

Would it make sense to just use this method?

nalimilan replied:
Unfortunately, it triggers the same test failure. But it would indeed make sense to define that method, which is more general.

vtjnash replied:
You can add that method to the list of allowed exceptions, since the necessary ambiguity resolver is present (we just have limited logic implemented there to compute and detect those intersections).

nalimilan commented on base/precompile.jl:
```diff
@@ -1115,28 +1115,28 @@ precompile(Tuple{typeof(Base._jl_spawn), String, Array{String, 1}, Ptr{Void}, Ba
-precompile(Tuple{Type{Base.Nullable{Array{T, N} where N where T}}})
-precompile(Tuple{typeof(Base.convert), Type{Base.Nullable{Base.Process}}, Base.Process})
-precompile(Tuple{typeof(Base.convert), Type{Base.Nullable{Bool}}, Bool})
+precompile(Tuple{typeof(Base.eltype), Type{Base.Union{Some{IO}, Null}}})
```
Does it make any sense to call `precompile` with `Union` types? Or should each line be duplicated, once for `Null` and once for `Some{T}`?

nalimilan commented:
> While it's the current name, I don't think get is the right name. Every other method of get require a default value to be passed.

I share that feeling, but I'd rather defer this discussion to another issue so that we keep this one focused. Maybe file a separate issue?

nalimilan commented:
I've pushed a new commit which gets rid of `Null`, `null`, `isnull` and `NullException`, and uses `Void`, `nothing` and `=== nothing` instead. As discussed at https://github.com/JuliaData/Roadmap.jl/issues/3#issuecomment-332763562, this allows keeping `Null` in Nulls.jl so that no representation of missing data is set in stone in Julia 1.0.

Any remaining comments on the design?

rfourquet commented:
I made an experiment in yet-unmerged #22324 to use `Some{T}` instead of `Nullable{T}`, and I must say I rather like this new API. In particular, I find using `nothing` for the missing value rather elegant. For example, `Nullable(value, pred)` can simply be replaced by `if pred value end`, or `for x in a; ... return Nullable(y)... end; return Nullable{K}()` can be replaced by `for x in a; ... return Some(y)... end`. 
Also, I wonder if we would want using `x[]` instead (or in addition to) `get(x)`, as `x[]` seems to also be a way to retrieve a wrapped value (in `Ref` or 0-dim arrays), meaning there is no need in some cases to require `Union{Some{T}, Void}`, but only `Union{X, Void}` where `X` supports indexing.
On the negative side, I could observe a performance degradation of about 30% (I may have made mistakes though, or missed optimization; I will double check).

nalimilan commented:
> Also, I wonder if we would want using x[] instead (or in addition to) get(x), as x[] seems to also be a way to retrieve a wrapped value (in Ref or 0-dim arrays), meaning there is no need in some cases to require Union{Some{T}, Void}, but only Union{X, Void} where X supports indexing.

It would indeed make sense to support `x[]`. But I don't see how this relates to whether `X` supports indexing or not.

> On the negative side, I could observe a performance degradation of about 30% (I may have made mistakes though, or missed optimization; I will double check).

Yes, not all optimizations for `Union` have been merged yet. Many uses of `Nullable` are not in performance-critical code (indeed, some previously used `nothing` before `Nullable` was introduced), but in some cases it could be a problem. I see it as an incentive to improve the compiler, which is good thing since we absolutely need this to be efficient for missing data too. ;-)

quinnj commented:
@rfourquet, the performance issues you saw were probably #23338.

rfourquet commented:
> But I don't see how this relates to whether X supports indexing or not.

I was unclear: I just meat "objects of type `X` support indexing (without indices)", i.e. `x[]` is defined when `x isa X`.

nalimilan commented:
I think the PR is now ready for a final review. I've added a temporary commit reverting #24197 since the tests fail without it due to a crash; it will have to be removed before merging.

nalimilan commented:
I've been able to drop the temporary revert commit since the crash has been fixed. Any chance to get this reviewed soon? The PR tends to accumulate non-trivial conflicts in many different areas quite quickly since it touches a lot of code.

vtjnash commented on base/libgit2/rebase.jl:
```diff
@@ -1,14 +1,14 @@
 
 function GitRebase(repo::GitRepo, branch::GitAnnotated, upstream::GitAnnotated;
-                   onto::Nullable{GitAnnotated}=Nullable{GitAnnotated}(),
+                   onto::Union{Some{GitAnnotated}, Void}=nothing,
```
Some wrapper seems unnecessary here

vtjnash commented on base/libgit2/reference.jl:
```diff
@@ -255,23 +255,24 @@ function lookup_branch(repo::GitRepo,
         throw(Error.GitError(err))
     end
-    return Nullable{GitReference}(GitReference(repo, ref_ptr_ptr[]))
+    return Some(GitReference(repo, ref_ptr_ptr[]))
```
wrapper is unnecessary

vtjnash commented on base/libgit2/reference.jl:
```diff
@@ -281,14 +282,14 @@ function upstream(ref::GitReference)
         throw(Error.GitError(err))
     end
-    return Nullable{GitReference}(GitReference(ref.owner, ref_ptr_ptr[]))
+    return Some(GitReference(ref.owner, ref_ptr_ptr[]))
```
wrapper is unnecessary

vtjnash commented on base/libgit2/remote.jl:
```diff
@@ -80,9 +81,9 @@ function lookup_remote(repo::GitRepo, remote_name::AbstractString)
                 rmt_ptr_ptr, repo.ptr, remote_name)
     if err == Int(Error.GIT_OK)
-        return Nullable{GitRemote}(GitRemote(repo, rmt_ptr_ptr[]))
+        return Some(GitRemote(repo, rmt_ptr_ptr[]))
```
wrapper is unnecessary

vtjnash commented on base/libgit2/repository.jl:
```diff
@@ -336,18 +336,18 @@ function checkout_tree(repo::GitRepo, obj::GitObject;
 See [`CheckoutOptions`](@ref) for more information.
 """
-function checkout_index(repo::GitRepo, idx::Nullable{GitIndex} = Nullable{GitIndex}();
+function checkout_index(repo::GitRepo, idx::Union{Some{GitIndex}, Void} = nothing;
```
wrapper is unnecessary

vtjnash commented on base/libgit2/repository.jl:
```diff
@@ -385,11 +385,11 @@ function cherrypick(repo::GitRepo, commit::GitCommit; options::CherrypickOptions
 
 """Updates some entries, determined by the `pathspecs`, in the index from the target commit tree."""
-function reset!(repo::GitRepo, obj::Nullable{<:GitObject}, pathspecs::AbstractString...)
+function reset!(repo::GitRepo, obj::Union{Some{<:GitObject}, Void}, pathspecs::AbstractString...)
```
wrapper is unnecessary

vtjnash commented on base/libgit2/status.jl:
```diff
@@ -45,6 +45,6 @@ function status(repo::GitRepo, path::String)
-    (ret == Cint(Error.ENOTFOUND) || ret == Cint(Error.EAMBIGUOUS)) && return Nullable{Cuint}()
-    return Nullable(status_ptr[])
+    (ret == Cint(Error.ENOTFOUND) || ret == Cint(Error.EAMBIGUOUS)) && return nothing
+    return Some(status_ptr[])
```
wrapper is unnecessary

vtjnash commented on base/libgit2/types.jl:
```diff
@@ -963,11 +963,11 @@ for (typ, owntyp, sup, cname) in [
-                    $typ($owntyp(owner), ptr, fin)
+                $typ(ptr::Ptr{Void}, fin::Bool=true) = $typ(nothing, ptr, fin)
+                $typ(owner::$(owntyp.args[2].args[2]), ptr::Ptr{Void}, fin::Bool=true) =
+                    $typ(Some(owner), ptr, fin)
```
The Some wrapper seems unnecessary. That'll also remove the need for the second method here too.

vtjnash commented on base/libgit2/types.jl:
```diff
@@ -1253,16 +1253,16 @@ A `CredentialPayload` instance is expected to be `reset!` whenever it will be us
 
     # Ephemeral state fields
-    credential::Nullable{AbstractCredentials}
+    credential::Union{Some{<:AbstractCredentials}, Void}
```
the Some wrappers aren't necessary in this type's fields

vtjnash commented on base/libgit2/types.jl:
```diff
@@ -1288,11 +1288,11 @@ mutable struct CredentialPayload <: Payload
 function CredentialPayload(credential::AbstractCredentials; kwargs...)
-    CredentialPayload(Nullable(credential), Nullable{CachedCredentials}(); kwargs...)
+    CredentialPayload(Some(credential), nothing; kwargs...)
 end
```
method can be deleted along with Some wrappers

vtjnash commented on base/lock.jl:
```diff
@@ -11,7 +11,7 @@ Each `lock` must be matched with an `unlock`.
 """
 mutable struct ReentrantLock
-    locked_by::Nullable{Task}
+    locked_by::Union{Some{Task}, Void}
```
Doesn't need the `Some` wrapper

vtjnash commented on base/mpfr.jl:
```diff
@@ -127,7 +127,7 @@ convert(::Type{BigFloat}, x::Rational) = BigFloat(numerator(x)) / BigFloat(denom
     z = BigFloat()
     err = ccall((:mpfr_set_str, :libmpfr), Int32, (Ref{BigFloat}, Cstring, Int32, Int32), z, s, base, ROUNDING_MODE[])
-    err == 0 ? Nullable(z) : Nullable{BigFloat}()
+    err == 0 ? Some(z) : nothing
```
no Some wrapper needed

vtjnash commented on base/parse.jl:
```diff
@@ -218,21 +217,36 @@ end
-    return unsafe_get(result)
+    return get(result)
 end
 
```
The `Some` wrappers in this file don't seem necessary

ararslan replied:
I disagree; the `Some` wrappers signal that the result was indeed parsed correctly.

vtjnash replied:
The value itself also signals that. There's no additional information being provided by the `Some` wrapper, just extra typing.

nalimilan replied:
There's a single case where `Some` would be needed: it's for `parse(Void, s)`. If we don't plan to support this (doesn't make a lot of sense), then I guess we can remove it.

Then one can always make the argument that `Some` forces the caller to handle the possibility of the result being `nothing`. There are advocates of this approach in other languages, though I'm not completely convinced bothering users is useful.

vtjnash replied:
> forces the caller to handle the possibility of the result being nothing

Basic test philosophy will also do that. If your code didn't need tests, then it probably also didn't care that much about handling failures either. Also, this is pretty easy to detect as a Lint pass. Also, it's called `tryparse`: it should be pretty obvious to a code-reviewer if you handle it like `parse`. Anyways, I'm aware that there's vast bodies of discussion about this, which I'm neither trying to dismiss or assert here, just stating what's been JuliaLang's typical stance on the matter, since the other approaches do indeed have their merits too.

StefanKarpinski replied:
I have to wonder if `tryparse` and `parse` shouldn't be merged with this change.

nalimilan replied:
> I have to wonder if tryparse and parse shouldn't be merged with this change.

Maybe, but let's defer this decision to another issue?

nalimilan replied:
Actually there are less absurd situations where `Some` might be needed. One might want to have a method like `tryparse(Union{Int, Void}, str)`, which should return `Some(nothing)` when the input is `"nothing"`, and `nothing` when the input is invalid. See https://github.com/JuliaLang/julia/pull/16981.

This situation is less likely now that we no longer conflate the engineer's null and the data scientist's null in a single type as `Nullable` did: we could just state that we don't care about allowing to parse `nothing`. OTOH there may be contexts where it makes sense to parse `"nothing"`, e.g. when parsing JSON files.

nalimilan replied:
After considering this question again, I've decided to return `Union{T, Void}` rather than `Union{Some{T}, Void}`. Indeed, the only case when the `Some` wrapper would be useful would be (as in the example above) when passing a `Union` type to `parse`. But that does not sound like an interesting use case. The only reason I can imagine in favor of supporting `parse(Union{T, U}}, s}` would be performance, if the parsing work can be share between the types. Else you can do `x = tryparse(T, s); x === nothing && (x = tryparse(U, s)`. And that performance hypothesis doesn't sound too realistic to me. For `tryparse(Union{T, Void}, s)`, you'd just check `s == "nothing"` or something like that, which is indeed what code does currently.

So I've added a commit to change this.

vtjnash commented on base/pkg/entry.jl:
```diff
@@ -310,15 +310,15 @@ function pin(pkg::AbstractString, head::AbstractString)
                 else
                     info("Creating $pkg branch $branch")
-                    ref = Nullable(LibGit2.create_branch(repo, branch, commit))
+                    ref = Some(LibGit2.create_branch(repo, branch, commit))
```
wrapper not needed

vtjnash commented on base/pkg/pkg.jl:
```diff
@@ -22,12 +22,12 @@ const META_BRANCH = "metadata-v2"
 struct PkgError <: Exception
     msg::AbstractString
-    ex::Nullable{Exception}
+    ex::Union{Some{<:Exception}, Void}
```
wrapper not needed

vtjnash commented on base/precompile.jl:
```diff
@@ -1114,28 +1114,28 @@ precompile(Tuple{typeof(Base._jl_spawn), String, Array{String, 1}, Ptr{Void}, Ba
-precompile(Tuple{Type{Base.Nullable{Array{T, N} where N where T}}})
-precompile(Tuple{typeof(Base.convert), Type{Base.Nullable{Base.Process}}, Base.Process})
-precompile(Tuple{typeof(Base.convert), Type{Base.Nullable{Bool}}, Bool})
+precompile(Tuple{typeof(Base.eltype), Type{Base.Union{Some{IO}, Void}}})
```
They usually need to be separate. Although note that in the case it's passing `Type{Union{...}}`, so that's one value, not a Union of types, so it is correct as-is

vtjnash commented on base/process.jl:
```diff
@@ -391,20 +391,20 @@ function _uv_hook_close(proc::Process)
 end
 
-function spawn(redirect::CmdRedirect, stdios::StdIOSet; chain::Nullable{ProcessChain}=Nullable{ProcessChain}())
+function spawn(redirect::CmdRedirect, stdios::StdIOSet; chain::Union{Some{ProcessChain}, Void}=nothing)
```
Some wrapper not needed

vtjnash commented on base/socket.jl:
```diff
@@ -262,7 +262,7 @@ mutable struct TCPSocket <: LibuvStream
     connectnotify::Condition
     closenotify::Condition
-    sendbuf::Nullable{IOBuffer}
+    sendbuf::Union{Some{IOBuffer}, Void}
```
`Some` wrapper not needed

vtjnash commented on base/stream.jl:
```diff
@@ -104,7 +104,7 @@ mutable struct PipeEndpoint <: LibuvStream
     connectnotify::Condition
     closenotify::Condition
-    sendbuf::Nullable{IOBuffer}
+    sendbuf::Union{Some{IOBuffer}, Void}
```
don't need wrapper here

vtjnash commented on base/stream.jl:
```diff
@@ -154,7 +154,7 @@ mutable struct TTY <: LibuvStream
     readnotify::Condition
     closenotify::Condition
-    sendbuf::Nullable{IOBuffer}
+    sendbuf::Union{Some{IOBuffer}, Void}
```
don't need wrapper here

vtjnash commented on base/threadcall.jl:
```diff
@@ -1,7 +1,7 @@
 
 const max_ccall_threads = parse(Int, get(ENV, "UV_THREADPOOL_SIZE", "4"))
-const thread_notifiers = [Nullable{Condition}() for i in 1:max_ccall_threads]
+const thread_notifiers = fill!(Vector{Union{Some{Condition}, Void}}(max_ccall_threads), nothing)
```
don't need `Some` wrapper

vtjnash commented on doc/src/manual/style-guide.md:
```diff
@@ -139,8 +139,6 @@ some alternatives to consider:
   * Determine whether there is a simple rule for when `x` is `nothing`. For example, often the field
     will start as `nothing` but get initialized at some well-defined point. In that case, consider
     leaving it undefined at first.
-  * If `x` really needs to hold no value at some times, define it as `::Nullable{T}` instead, as this
```
The section is still potentially valid as a design recommendation ("simple is better than complex"). But it seems we usually ignore the "store them in a dictionary" proposal, so the only useful suggestion might be "Find a safe default value, or consider starting with it being uninitialized entirely". So I agree that this section is not all that helpful and can be removed.

vtjnash commented on test/distributed_exec.jl:
```diff
@@ -258,8 +258,8 @@ function test_regular_io_ser(ref::Base.Distributed.AbstractRemoteRef)
-            @test v === Nullable{Any}()
+        elseif isa(v, Union{Some, Void})
+            @test v === nothing
         else
```
This seems like a strange way to write it. Perhaps instead make the error branch an elseif and remove this `@test`?
```
elseif v !== nothing
```

vtjnash commented on test/libgit2-helpers.jl:
```diff
@@ -12,7 +12,7 @@ without having to authenticate against a real server.
         valid_credential::AbstractCredentials,
         url::AbstractString,
-        user::Nullable{<:AbstractString},
+        user::Union{Some{<:AbstractString}, Void},
```
`Some` wrappers not necessary in this file

vtjnash commented on base/some.jl:
```diff
@@ -0,0 +1,68 @@
+
+get(x::Some) = x.value
+get(x::Some, y) = x.value
+get(x::Void, y) = y
```
also implement `x[]`?
```julia
getindex(x::Some) = x.value
```

nalimilan replied:
That's been suggested in the comments, but I'd rather leave this for another PR as that's a feature that can be added at any point, and it opens a broad discussion regarding whether `Some` is a container or not.

vtjnash replied:
OK, that's fine

vtjnash commented:
I may have been a bit thorough at marking all places where `Union{T, Void}` would be easier to use :P

Anyways, looking good! Thanks for tackling this!

ararslan commented on base/deprecated.jl:
```diff
@@ -2093,6 +2093,11 @@ end
 
+@deprecate_moved Nullable "Nullables"
+@deprecate_moved NullException "Nullables"
+@deprecate_moved isnull "Nullables"
```
I think it's safer to deprecate the specific method for `isnull(::Nullable)`, since `isnull` will be used for `Null`.

nalimilan replied:
According to the current plans, `isnull` will be defined in Nulls, which will be either a plain package or a part of the standard library, right? So we need to deprecate the function itself in Base, don't we?

ararslan commented on base/deprecated.jl:
```diff
@@ -2093,6 +2093,11 @@ end
+@deprecate_moved Nullable "Nullables"
+@deprecate_moved NullException "Nullables"
+@deprecate_moved isnull "Nullables"
+@deprecate_moved unsafe_get "Nullables"
```
Same here. Was `unsafe_get` only used for `Nullable`?

nalimilan replied:
Yes.

nalimilan commented:
> I may have been a bit thorough at marking all places where Union{T, Void} would be easier to use :P

Thanks. My plan was to replace (almost) all uses of `Nullable{T}` with `Union{Some{T}, Void}` first, and then improve the code on a case-by-case basis in subsequent PRs. That would avoid losing the important design discussions in a mountain of comments and changes. What do you think?

nalimilan commented on base/dates/io.jl:
```diff
@@ -19,7 +19,8 @@ the argument out in the method definition.
 Returns a tuple of 2 elements `(res, idx)`, where:
 
-* `res` is a `Nullable{T}` - the result of the parsing, null if parsing failed.
+* `res` is either a [`Some`](@ref) object holding the result of the parsing,
```
I don't think that would work, since when it fails the index is used to point at the place where parsing ended.

vtjnash replied:
I don't think any of the code actually used that though (it just either threw an error or used the position from before the call). I think this was just a side-effect of working extra hard to make it type-stable.

nalimilan commented on base/deprecated.jl:
```diff
@@ -1722,20 +1722,20 @@ import .Iterators.enumerate
 # `payload::CredentialPayload` from base/libgit2/libgit2.jl
 @eval LibGit2 function deprecate_nullable_creds(f, sig, payload)
-    if isa(payload, Nullable{<:AbstractCredentials})
+    if isa(payload, Union{Some{<:AbstractCredentials}, Void})
```
This one is a bit special. This function is called by internal code, so it needs to work with the new framework. Else we would break even more things than needed.

nalimilan commented on base/distributed/remotecall.jl:
```diff
@@ -16,10 +16,11 @@ mutable struct Future <: AbstractRemoteRef
-    Future(w::Int, rrid::RRID) = Future(w, rrid, Nullable{Any}())
-    Future(w::Int, rrid::RRID, v) = (r = new(w,rrid.whence,rrid.id,v); return test_existing_ref(r))
+    Future(w::Int, rrid::RRID) = Future(w, rrid, nothing)
+    Future(w::Int, rrid::RRID, v::Union{Some{<:Any}, Void}) =
```
Yeah, maybe. Though I generally like that you can pass `nothing` as a second argument, this can sometimes be cleaner than calling either the two- or three-argument version depending on whether you want to set the value.

vtjnash replied:
Sure. I would just rather see it written as:
```
Future(w::Int, rrid::RRID, v::Void=nothing) = Future(w, rrid, v)
Future(w::Int, rrid::RRID, v) = Future(w, rrid, Some(v))
```
So that the code is clear about its intentions.

Is it that common to have code that maybe-sets the value?

nalimilan replied:
> Is it that common to have code that maybe-sets the value?

No idea, but I know it's hard to write generic code when you can't just pass the values to functions, but instead need to change the way the function is invoked (I suffered from this in R, which offers a `missing` function to check whether an argument was left to its default value, which isn't equivalent to explicitly passing the default value).

nalimilan replied:
Actually the two methods proposed above are problematic since that implies `Future(w, rrid, 1)` would store `Some(1)` but `Future(w, rrid, nothing)` would store `nothing` (and not `Some(nothing)`), which defeats the purpose of using `Some`. That's the same problem as allowing `convert(Some, 1)` to work. So we have to require `v::Union{Some{<:Any}, Void}`.

nalimilan commented on base/docs/basedocs.jl:
```diff
@@ -810,7 +810,7 @@ Void
 
 The singleton instance of type `Void`, used by convention when there is no value to return
-(as in a C `void` function). Can be converted to an empty `Nullable` value.
+(as in a C `void` function) or when a variable or field holds no value.
```
"not available" sounds a bit too much like R's `NA`, of which `nothing` isn't the equivalent. Also it's not only for return values, which is why I mentioned variables and fields.

StefanKarpinski replied:
In the future once `missing` is standard, we could mention it here by comparison:
* `nothing` means "there is no value"
* `missing` means "there is a value but we don't have it"

nalimilan commented on base/deprecated.jl:
```diff
@@ -2093,6 +2093,11 @@ end
     @deprecate chol!(x::Number, uplo) chol(x) false
 end
 
+@deprecate_moved Nullable "Nullables"
```
I wonder whether it wouldn't make more sense to deprecate `Nullable` in favor of `Union{Some{T}, Void}` instead. `isnull` would have to be deprecated in favor of `x === nothing`. The problem is that it's not a perfect replacement for `Nullable` so lots of things would break. But it would work for places where people call `isnull` on a value returned e.g. by `tryparse`, when they will currently get a rather misleading message about using `Nullables`.

StefanKarpinski commented on base/channels.jl:
```diff
@@ -18,13 +18,13 @@ Other constructors:
+    cond_put::Condition                  # waiting for a writeable slot
     state::Symbol
-    excp::Nullable{Exception} # Exception to be thrown when state != :open
+    excp::Union{Some{<:Exception}, Void} # Exception to be thrown when state != :open
```
To elaborate: a `Some` wrapper is only needed when `Void <: T`; in this case since `Void` is not a subtype of `Exception`.

vtjnash commented:
I think we should go ahead and use `Union{T, Void}` as much as possible, but have `Union{Some{T}, Void}` available for the cases where it is not (an example from this PR of such is `Future`, where adding an explicit extra `isset::Bool` field would likely be more annoying)

nalimilan commented:
OK, I'll try to use `Union{T, Void}` everywhere it makes sense and it's not controversial (keeping in mind we can always make more changes later).

nalimilan committed:
Use Union{T, Void} rather than Union{Some{T}, Void} where possible

nalimilan commented:
I've added a commit switching to `Union{T, Void}` for all cases noted by @vtjnash (with even two more cases which made sense), except for `tryparse` which is more complex. Overall this (of course) makes the code simpler, and `Some` is used in very few places. But there are two issues to discuss:
- In situations where `get(x, default)` was used, the code is now less clear: it needs to become `x === nothing ? default : x`, which can be ugly when `x` is long. We should find a solution for this, e.g. via `coalesce(::Void, default::Any) = default` and `coalesce(x::Any, ::Any) = x`, and possibly `x ?? default` (null-coalescing operator, which can be added after 1.0 if we make sure the syntax is an error now). Better not keep using `get` for this since we're no longer able to dispatch on a specific type like `Nullable`, so it could be confusing. `coalesce` will be useful for `missing` too (it's already defined in Nulls, mirroring SQL's function with the same name).
- Uses of `get` ensured that an unhandled `nothing` would trigger a failure early. This is the classic argument in favor of forcing unwrapping. In some complex code paths, it's not clear whether an unhandled `nothing` would trigger an error or succeed silently (e.g. `x == y` when both values could be `nothing`), so being able to concisely add assertions that a value is not `nothing` would be nice. Again, it could be a simple function like `checknothing(x) = x` with `checknothing(::Void) = throw(MethodError())` (see https://github.com/JuliaDocs/Documenter.jl/pull/576), but it could make sense to have a special syntax like `x::!Void`.

Finally, we need to decide what to do for `tryparse`: see [above](https://github.com/JuliaLang/julia/pull/23642#discussion_r150101238).

nalimilan committed:
Update precompile.jl

Remove all uses of Some, except for Union{Some{Any}, Void}.

ViralBShah commented:
Bump. What needs to be done here to merge? This would be huge to do, and bumping in order to figure out whose time we need before feature freeze.

nalimilan commented:
I think this is ready to merge, though we need to discuss the `tryparse` issue either before or after merging. The two points I mentioned above can be tackled later.

quinnj reviewed (approved):
Did `Nullables.jl` get registered/updated yet? Also make sure we squash when merging this.

nalimilan commented:
Bump! Can we merge this? It has accumulated lots of conflicts again, and even more are coming with https://github.com/JuliaLang/julia/pull/24738...

rfourquet reviewed (approved):
Superficially looks good to me, modulo maybe few changes suggested in comments. ("superficially" in that I didn't check everything is correct (which would probably take me days), I just checked that it looks correct, and didn't review broadcast.jl) 
My main gripe is that we need a function to handle the pattern `x == nothing : default : x`, e.g. `get(x, default)`, in particular when `x` is an expression to store in a temporary variable. But I guess this is not blocking and can be added any time later.

nalimilan commented on base/dates/parse.jl:
```diff
@@ -24,12 +24,13 @@ genvar(t::DataType) = Symbol(lowercase(string(Base.datatype_name(t))))
-* `values::Nullable{Tuple}`: A tuple which contains a value for each `DatePart` within the
-  `DateFormat` in the order in which they occur. If the string ends before we finish parsing
-  all the directives the missing values will be filled in with default values.
+* `values::Union{Some{Tuple}, Void}`: Either `nothing`, or a `Some` object wrapping a tuple
```
That's because of the general issue with `tryparse`, see my comments above.

nalimilan commented on base/client.jl:
```diff
@@ -70,7 +70,7 @@ color_normal = text_colors[:normal]
     env_str = get(ENV, key, "")
     c = tryparse(Int, env_str)
-    c_conv = isnull(c) ? Symbol(env_str) : get(c)
+    c_conv = c === nothing ? Symbol(env_str) : get(c)
```
Sure, that's just because I took the existing code without thinking too much about how to improve it.

nalimilan commented on base/deprecated.jl:
```diff
@@ -1571,8 +1571,9 @@ end
             "`LibGit2.prompt(msg::AbstractString; default::AbstractString=\"\", password::Bool=false)` is deprecated, use ",
-            "`get(Base.prompt(msg, default=default, password=password), \"\")` instead."), :prompt)
-        Base.get(Base.prompt(msg, default=default, password=password), "")
+            "`result = Base.prompt(msg, default=default, password=password); result === nothing ? \"\" : result` instead."), :prompt)
```
Yes, as I noted above it would make sense to have a convenience function for that (not sure what it would be exactly). But it can always be added later (contrary to API breaks).

rfourquet replied:
Sorry, I forgot to reread those last comments in this thread! If no good name is found before this gets merged, I think I would be in favor of using a non-exported temporary name, like `_get` or whatever, so that at least the pattern is captured (and easily replaced when the new name comes), and at least something is available for the code which will be written in the meantime.

nalimilan commented on base/distributed/cluster.jl:
```diff
@@ -174,15 +172,13 @@ worker_timeout() = parse(Float64, get(ENV, "JULIA_WORKER_TIMEOUT", "60.0"))
 worker.
 
-If the cookie is unspecified, the worker tries to read it from its STDIN.
+If the cookie is `nothing`, the worker tries to read it from its STDIN.
```
Again, I guess I just took the existing code without thinking too much about why it used `Nullable`.

rfourquet replied:
Then maybe better to leave "unspecified" here, I see no reason make this documentation change and it's still possible to keep it unspecified after your change.

nalimilan commented on base/distributed/cluster.jl:
```diff
@@ -455,9 +445,9 @@ function setup_launched_worker(manager, wconfig, launched_q)
     # process on the remote machine, with a request to start additional workers of the
     # same type. This is done by setting an appropriate value to `WorkerConfig.cnt`.
-    cnt = get(wconfig.count, 1)
+    cnt = wconfig.count === nothing ? 1 : wconfig.count
```
See my comments above.

nalimilan commented on base/distributed/cluster.jl:
```diff
@@ -550,10 +540,13 @@ function create_worker(manager, wconfig)
+        filterfunc(x) = (x.id != 1) && isdefined(x, :config) &&
+            (x.config.ident !== nothing) && (wconfig.connect_idents !== nothing) &&
+            (x.config.ident in wconfig.connect_idents)
 
```
Yes, that's one of the annoying consequences of the change that I've highlighted above. We would need some kind of type assertion, or a function, to check that the value is not `nothing`.

nalimilan commented on base/distributed/remotecall.jl:
```diff
@@ -16,12 +16,12 @@ mutable struct Future <: AbstractRemoteRef
 
-    Future(w::Int, rrid::RRID) = Future(w, rrid, Nullable{Any}())
-    Future(w::Int, rrid::RRID, v) = (r = new(w,rrid.whence,rrid.id,v); return test_existing_ref(r))
+    Future(w::Int, rrid::RRID, v::Union{Some{<:Any}, Void}=nothing) =
```
Right, `<:Any` can be dropped. Conversion of `Some{Int}` to `Some{Any}` happens automatically.

nalimilan commented on base/libgit2/libgit2.jl:
```diff
@@ -304,8 +304,8 @@ function push(repo::GitRepo; remote::AbstractString="origin",
-              payload::Union{CredentialPayload,Nullable{<:AbstractCredentials}}=CredentialPayload())
-    p = reset!(deprecate_nullable_creds(:push, "repo", payload), GitConfig(repo))
+              payload::Union{CredentialPayload,AbstractCredentials,Void}=CredentialPayload())
+    p = reset!(deprecate_nullable_creds(:push, "repo", payload))
```
Good catch! I don't think so, I probably messed up something during a rebase. And since deprecated functions are not tested...

nalimilan commented on base/some.jl:
```diff
@@ -0,0 +1,68 @@
+
+A wrapper type used with `Union{Some{T}, Void}` to distinguish between the absence
+of a value ([`nothing`](@ref)) and the presence of a `nothing` value (i.e. `Some(nothing)`).
+It can also be used to force users of a function argument or of an object field
```
Right. I wonder whether I shouldn't remove this sentence anyway, since we don't encourage people to force manual unwrapping. APIs would be more consistent if `Some` was only used where `nothing` is a possible valid value.

nalimilan commented on base/some.jl:
```diff
@@ -0,0 +1,68 @@
+    value::T
+end
+
+eltype(::Type{Some{T}}) where {T} = T
```
I guess I can remove it for now, that would mostly makes sense if we supported `getindex`.

nalimilan commented on base/some.jl:
```diff
@@ -0,0 +1,68 @@
+promote_rule(::Type{Some{T}}, ::Type{Void}) where {T} = Union{Some{T}, Void}
+
+convert(::Type{Some{T}}, x::Some) where {T} = Some{T}(convert(T, x.value))
+convert(::Type{Union{Some{T}, Void}}, x::Some) where {T} = convert(Some{T}, x)
```
Not completely, as `Nullable` currently supports more conversions, e.g. `convert(Nullable{Int}, nothing) === Nullable{Int}()`, which introduces an ambiguity about whether `nothing` is a valid value or a missing value. IIRC this method is needed so that you can assign e.g. `Some(1)` to a `Union{Some{Real}, Void}` field and get the automatic conversion.

nalimilan commented on base/some.jl:
```diff
@@ -0,0 +1,68 @@
+5
+
+julia> get(nothing)
+ERROR: MethodError()
```
Yes, but I was thinking we should use a different method than `get` for this to make it clearer. I guess we could introduce `VoidException`.

nalimilan commented on test/core.jl:
```diff
@@ -3234,17 +3234,21 @@ f11858(Any[Type{Foo11858}, Type{Bar11858}, typeof(g11858)])
 # issue 11904
+struct Nullable11904{T}
+    value::T
+    hasvalue::Bool
```
Yeah, but it's simpler that way since we don't need a custom constructor.

nalimilan commented on test/core.jl:
```diff
@@ -3468,7 +3472,7 @@ call13007(::Type{Array}) = 1
 
 # detecting cycles during type intersection, e.g. #1631
-cycle_in_solve_tvar_constraints(::Type{Nullable{S}}, x::S) where {S} = 0
+cycle_in_solve_tvar_constraints(::Type{Some{S}}, x::S) where {S} = 0
```
Because AFAICT all we need is a simple parametric type.

nalimilan commented on test/serialize.jl:
```diff
@@ -357,11 +357,17 @@ create_serialization_stream() do s # user-defined type array
 end
 
 # corner case: undefined inside immutable struct
+struct MyNullable{T}
```
These are really not related, `Nullable` was only used as a way to trigger a particular bug. On the contrary, it's better to be sure the type isn't modified, which could stop exposing the bug. We could even rename them to `CustomStruct3759`.

KristofferC commented:
This needs a nanosoldier run before begin merged I presume.

rfourquet commented:
Should we run nanosoldier here? (but it seems to be temporarily unavailable) 
Despite the efficiency improvements of `Union`, I can still observe performance degradation in some benchmarks, sometimes by 1.5x or 2x... so I'm a bit worried that this change gives incentive to avoid using what was `Nullable` in performance sensitive code. Are more improvements expected on `Union`s ?

nalimilan commented:
Running Nanosoldier is always a good idea, but unfortunately if we notice regressions we don't have a good alternative right now, except improving the compiler (now or later). See https://github.com/JuliaLang/julia/issues/23336 and https://github.com/JuliaLang/julia/issues/23338 about improving `Union` performance.

KristofferC commented:
We should still run it to see that the changes are at least within expected range.

rfourquet commented:
In principle, is it likely to have `Union{T,Void}` as efficient as `Nullable` eventually? otherwise `Nullable` will have to continue to exist in some form or another (in a package presumably).

nalimilan commented:
> In principle, is it likely to have Union{T,Void} as efficient as Nullable eventually? otherwise Nullable will have to continue to exist in some form or another (in a package presumably).

AFAIK there's no intrinsic reason why it couldn't be made as efficient, but I'm not sure how much work it will require.

ararslan commented:
Nanosoldier should be back up soon. In the meantime, this PR should be rebased against master or else Nanosoldier will check out the branch tip rather than the merge commit with master.

nalimilan committed:
Review fixes

nalimilan commented on test/broadcast.jl:
```diff
@@ -609,4 +606,4 @@ end
 
 # Issue #22180
-@test isequal(convert.(Nullable, [1,2]), [Nullable(1), Nullable(2)])
+@test convert.(Any, [1, 2]) == [1, 2]
```
@timholy Is this OK? I took the test from https://github.com/JuliaLang/julia/pull/22180, but I guess there was a reason why you changed it?

ararslan commented:
@nanosoldier `runbenchmarks(ALL, vs=":master")`

rfourquet commented:
Maybe nanosoldier didn't run because of broken tests (cf. #24800) ? re-trying: @nanosoldier `runbenchmarks(ALL, vs=":master")`

nalimilan commented:
Actually, that's because BaseBenchmarks still tests `Nullable`. See https://github.com/JuliaCI/BaseBenchmarks.jl/pull/144.

ararslan commented:
For whatever reason, Nanosoldier seems to be just straight up ignoring requests from this PR. If there was an actual error, it would still show a status on this PR. Unfortunately I can't see the webhook log for this repo since I'm not an admin, so I'm not sure how to diagnose properly. I'll restart the server and once Milan's BaseBenchmarks PR is merged I'll trigger another run here.

nalimilan committed:
Add coalesce() and use it

nalimilan committed:
Fix start_worker() method definition

nalimilan committed:
Add notnothing(x) to assert that x !== nothing

Use it everywhere get() was called and nothing would not trigger an error immediately.

nalimilan commented:
@nanosoldier `runbenchmarks(ALL, vs = ":master")`

nalimilan commented:
I have added a `coalesce(x, y)` function to replace ugly `x == nothing ? y : x` repetitions, and a `notnothing(x)` function used to assert that `x !== nothing` everywhere `nothing` could propagate without triggering an error immediately (which was previously checked by `get(x)`). For now they are not exported so that we can discuss the design, in particular in the context of `missing` (which should also work with `coalesce`).

So it should be good now, if only Nanosoldier accepted to run...

ararslan commented:
@nanosoldier `runbenchmarks(ALL, vs=":master")`

nanosoldier commented:
Something went wrong when running [your job](https://github.com/JuliaLang/julia/pull/23642#issuecomment-348317734):
```
NanosoldierError: failed to run benchmarks against primary commit: failed process: Process(`sudo cset shield -e su nanosoldier -- -c ./benchscript.sh`, ProcessExited(1)) [1]
```
Logs and partial data can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/23305bdec5731ec4938a104fa9c4ee224211ed1d/6a346e7_vs_93fc492/report.md)
cc @ararslan

ararslan commented:
Ah, JSON uses `Nullable`, so this kills JSON before it even gets to the benchmarks.

TotalVerb commented:
We can change JSON to use the Nullables.jl package for now, which should be an easy patch.

nalimilan commented:
@nanosoldier `runbenchmarks(ALL, vs=":master")`

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/23642#issuecomment-348755898) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/6994a819ea3d35701e57a2cf0bce3acb8fb8086b/82c58bb_vs_10ebd63/report.md). cc @ararslan

nalimilan commented:
Wow, lots of possible regressions. Some appear legitimate, like the "parse" tests for dates, but some are more suspicious. Let's try again to be sure: @nanosoldier `runbenchmarks(ALL, vs=":master")`

quinnj commented:
I imagine most of the regressions are https://github.com/JuliaLang/julia/issues/23338

nalimilan commented:
Probably. In the short term, if we want to avoid regressions in the dates parsing code, we could change the internal parsing functions to return `(value, isvalid)` rather than `isvalid ? nothing : value`. There's no reason why `parse` should suffer from `Union` performance penalties (only `tryparse` has to).

I'd appreciate if somebody could try running Nanosoldier again, it doesn't want to respond to my requests again.

omus commented:
@nanosoldier `runbenchmarks(ALL, vs=":master")`

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/23642#issuecomment-348947175) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/2e23091f59deb3042493d367ca648e6c04f9677f/525ef1d_vs_9b1a56e/report.md). cc @ararslan

nalimilan commented:
OK, here's the list of regressions present in both runs (some improvements are also present in both, but that's not possible):

| ID | time ratio | memory ratio |
|----|------------|--------------|
| `["array", "index", "(\"sumvector_view\", \"SubArray{Int32,2,Array{Int32,3},Tuple{Int64,Base.Slice{Base.OneTo{Int64}},Base.Slice{Base.OneTo{Int64}}},true}\")"]` | 1.66 (50%) :x: | 1.00 (1%)  |
| `["dates", "parse", "(\"DateTime\", \"DateFormat\")"]` | 1.35 (15%) :x: | 1.19 (1%) :x: |
| `["dates", "parse", "(\"DateTime\", \"ISODateTimeFormat\")"]` | 2.00 (15%) :x: | 12.00 (1%) :x: |
| `["dates", "parse", "(\"DateTime\", \"RFC1123Format\", \"Lowercase\")"]` | 3.90 (15%) :x: | 7.79 (1%) :x: |
| `["dates", "parse", "(\"DateTime\", \"RFC1123Format\", \"Mixedcase\")"]` | 3.33 (15%) :x: | 4.17 (1%) :x: |
| `["dates", "parse", "(\"DateTime\", \"RFC1123Format\", \"Titlecase\")"]` | 3.80 (15%) :x: | 7.79 (1%) :x: |
| `["dates", "parse", "(\"Date\", \"DateFormat\")"]` | 1.10 (15%)  | 1.16 (1%) :x: |
| `["dates", "parse", "(\"Date\", \"ISODateFormat\")"]` | 3.58 (15%) :x: | 23.50 (1%) :x: |
| `["dates", "parse", "Date"]` | 26.12 (15%) :x: | 23.50 (1%) :x: |
| `["dates", "parse", "DateTime"]` | 8.49 (15%) :x: | 12.00 (1%) :x: |
| `["misc", "parse", "DateTime"]` | 9.05 (15%) :x: | 11.95 (1%) :x: |
| `["parallel", "remotecall", "(\"identity\", 1024)"]` | 1.05 (15%)  | 1.09 (1%) :x: |
| `["parallel", "remotecall", "(\"identity\", 2)"]` | 1.05 (15%)  | 1.16 (1%) :x: |
| `["parallel", "remotecall", "(\"identity\", 4096)"]` | 1.06 (15%)  | 1.04 (1%) :x: |
| `["parallel", "remotecall", "(\"identity\", 512)"]` | 1.05 (15%)  | 1.12 (1%) :x: |
| `["parallel", "remotecall", "(\"identity\", 64)"]` | 1.04 (15%)  | 1.16 (1%) :x: |
| `["problem", "spellcheck", "spellcheck"]` | 2.52 (15%) :x: | 2.12 (1%) :x: |

Summary:
- "sumvector_view": I really don't see how it could be related, and the regression is just above the significance threshold (70% compared with 50%), without any increase in allocations. By calling `@btime` a few times locally, the difference does not appear to be consistent. So I think it is spurious.
- "remotecall" regressions don't sound too bad (+5%).
- Dates parsing regressions are terrible (even worse in the second run), but they are fixable by refactoring the code (as I noted above) if the compiler doesn't improve before the release.
- "spellcheck": this is one is also serious. It turns out it's due to `ProductIterator`, which uses `nothing` in the iteration state. I don't see how to fix it without introducing a `Nullable`-like type for internal use. I've tried using `Ref` instead, and indeed that reduces the number of allocations, but not down to what happens on master -- not sure why.

Overall the big question is: do we expect the compiler to improve in time or not?

mschauer commented:
The sentence "use  `Union{T, Void}` or `Union{Some{T}, Void}` instead" in the release notes begs the question "which?". Can we point to a place where the questios this raises are answered?

StefanKarpinski commented:
The answer is pretty simple: use `Some{T}` if it's possible that `Void <: T`, otherwise just use `T`.

StefanKarpinski commented:
cc @vtjnash – this needs your attention

nalimilan committed:
Fix ProductIterator regression by using custom MaybeValue wrapper

MaybeValue is just a simplified version of Nullable reserved for internal
use which wraps a value or no value in a type-stable way. Information
about whether a value is present or not is carried separately.

nalimilan committed:
Improve NEWS

nalimilan commented:
I've added an internal `MaybeValue` wrapper type to fix the performance regression in `ProductIterator`. Since it's only used as the iterator state, we can use whatever we want (we just need an object to hold a uninitialized value when there's no value to return).

@nanosoldier `runbenchmarks("problem", vs = ":master")`

maleadt commented:
@nanosoldier `runbenchmarks("problem", vs = ":master")`

andreasnoack commented:
@nanosoldier `runbenchmarks("problem", vs = ":master")`

ararslan commented:
@nanosoldier `runbenchmarks("problem", vs=":master")`

ararslan commented:
I've restarted the server. @nalimilan, can you try triggering Nanosoldier?

nalimilan commented:
@nanosoldier `runbenchmarks("problem", vs = ":master")`

ggggggggg commented:
Re: "The answer is pretty simple: use `Some{T}` if it's possible that `Void <: T`, otherwise just use `T`." More concretely: "Use `Union{T,Void}` in most cases, use `Union{Some{T},Void}` when `T=Void` or `T=Any`."?

StefanKarpinski commented:
There are many other types that include `nothing` – e.g. unions.

nalimilan committed:
Make tryparse() return Union{T, Void}

nalimilan committed:
Improve Some, remove get(::Some[, x]) method in favor of coalesce (now exported)

nalimilan committed:
LibGit2 rebase fixes

nalimilan commented:
I've rebased, improved NEWS, changed `tryparse` to return `Union{T, Void}` rather than `Union{Some{T}, Void}` as discussed on Slack (see [this comment](https://github.com/JuliaLang/julia/pull/23642#discussion_r150101238)), removed the `get` method for `Some` in favor of `coalesce`, which is now exported (and will be used for `Missing` too).

Please merge-squash if CI passes, since conflicts are very frequent and I'm not sure I'll survive another rebase. I'll do it if nobody beats me to it.

ararslan commented:
Looks like the change from `Union{Some{T},Void}` to `Union{T,Void}` may have introduced an ambiguity:
```
Test Failed at /tmp/julia/share/julia/test/ambiguous.jl:156
  Expression: detect_ambiguities(Core, Base; imported=true, recursive=true, ambiguous_bottom=false) == []
   Evaluated: Tuple{Method,Method}[(convert(::Type{Union{Missing, T}}, x) where T in Base at missing.jl:33, convert(::Type{Union{Void, T}}, x) where T in Base at some.jl:21)] == Any[]
```

nalimilan committed:
Test fixes

nalimilan committed:
Remove test/nullable.jl

nalimilan committed:
Fix remaining call to get() on Windows

nalimilan committed:
Support missing in coalesce()

rfourquet commented:
Alléluia :tada:

quinnj commented:
Fantastique!

ararslan commented:
Excellent work here, Milan, and thanks for sticking through so many rebases. You're a hero!

mlhetland commented:
There are two contradictory entries in [`NEWS.md`](https://github.com/JuliaLang/julia/blob/master/NEWS.md), with similar text except the instruction to use `Void` or `Nothing` as the type for `nothing`, and one mentioning the removal of `NullException`.

nalimilan commented:
Good catch @mlhetland. See https://github.com/JuliaLang/julia/pull/25573.
